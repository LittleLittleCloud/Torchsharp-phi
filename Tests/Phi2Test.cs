using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ApprovalTests;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static TorchSharp.torch;
using Xunit;
using TorchSharp;

namespace Phi.Tests;

public class Phi2Test
{
    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public async Task LoadSafeTensorShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\phi-2";
        var model = Phi2ForCasualLM.FromPretrained(modelWeightFolder, torchDtype: ScalarType.BFloat16, checkPointName: "model.safetensors.index.json");
        var state_dict_str = model.Model.Peek();
        Approvals.Verify(state_dict_str);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public async Task ForwardTest()
    {
        // create dummy input id with 128 length and attention mask
        var device = "cuda";
        var inputIds = torch.arange(128, dtype: ScalarType.Int64, device: device).unsqueeze(0);
        var attentionMask = torch.ones(1, 128, device: device);
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\phi-2";
        var model = Phi2ForCasualLM.FromPretrained(modelWeightFolder, torchDtype: ScalarType.BFloat16, checkPointName: "model.safetensors.index.json", device: "cuda");

        var (outputTokenIds, outputLogits) = model.Generate(inputIds, attentionMask);

        var outputTokenIdsStr = outputTokenIds.Peek("output");
        var outputLogitsStr = outputLogits.Peek("logits");

        var sb = new StringBuilder();
        sb.AppendLine(outputTokenIdsStr);
        sb.AppendLine(outputLogitsStr);

        Approvals.Verify(sb.ToString());
    }
}
