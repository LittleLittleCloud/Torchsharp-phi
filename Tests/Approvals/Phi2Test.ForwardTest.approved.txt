output: sum: 480.2512  dtype: BFloat16 shape: [1,128,2560]
logits: sum: 1711.1827  dtype: BFloat16 shape: [1,128,51200]
